### DISCUSSION TITLE
Headless browser / node.js app

### DISCUSSION THREAD

**Discussion by niutech** [emadda/tabserve#4] [Q&A]
Is it possible to run TabServe in a headless browser or as a Node.js app using `ws` library? The goal is to use it on a VPS without X11, like `localtunnel` or `ngrok`.

**Comment by emadda** [comment]
Using a Node ws to directly connect to the Cloudflare Worker is possible (you would not need the Tabserve web app).

The CF Worker receives HTTP requests then serialises them with msgpack over the websocket, so your Node function would use msgpack to parse the ws message into JSON and then use `fetch` to run the request.

This is where the CF Worker handles WS connections:
https://github.com/emadda/worker-tabserve-reverse-proxy/blob/master/src/index.ts#L143


You could also use a headless browser, but you would need to use CSS selectors and the DOM API to toggle the on/off buttons. A whole browser would probably be overkill if you can run some small tunnelling CLI. Take a look at https://github.com/anderspitman/awesome-tunneling
